blackbird_python/blackbird/utils.py

```diff
@@ -0,0 +1,245 @@
+# Copyright 2019 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# pylint: disable=too-many-return-statements,too-many-branches,too-many-instance-attributes
+"""
+.. _utils:
+
+Utilities
+=========
+
+**Module name:** `blackbird.utils`
+
+.. currentmodule:: blackbird.utils
+
+This module contains utility functions for Blackbird programs,
+including generating directed acyclic graphs, and pattern matching
+and Blackbird programs and templates.
+
+
+Summary
+-------
+
+.. autosummary::
+    TemplateError
+    to_DiGraph
+    match_template
+
+Code details
+~~~~~~~~~~~~
+"""
+from collections import namedtuple
+
+import numpy as np
+import sympy as sym
+from sympy.solvers import solve
+
+import networkx as nx
+from networkx.algorithms import isomorphism
+
+
+from .listener import RegRefTransform
+
+
+Command = namedtuple('Command', ['name', 'args', 'kwargs', 'modes'])
+"""namedtuple: encapsulate a specific quantum operation"""
+
+
+class TemplateError(Exception):
+    """Exception class for template related errors"""
+    pass
+
+
+def to_DiGraph(program):
+    """Convert a Blackbird program to a directed acyclic graph.
+
+    The resulting graph has nodes representing quantum operations,
+    and edges representing dependent/successor operations.
+
+    Each node is labelled by an integer; note attributes are used
+    to store information about the node:
+
+    * ``'name'`` *(str)*: name of the quantum operation (e.g., ``'S2gate'``)
+
+    * ``'args'`` *(list)*: positional arguments of the operation
+
+    * ``'kwargs'`` *(dict)*: keyword arguments of the operation
+
+    * ``'modes'`` *(tuple[int])*: modes the operation acts on
+
+    Args:
+        program (blackbird.Program): a Blackbird program
+
+    Returns:
+        networkx.DiGraph: the directed acyclic graph representing
+        the quantum program
+    """
+    grid = {}
+
+    for idx, op in enumerate(program.operations):
+        dependencies = set(op['modes'])
+
+        if 'args' in op:
+
+            for a in op['args']:
+                if isinstance(a, RegRefTransform):
+                    dependencies |= set(a.regrefs)

```

Comment:
whoa, never seen `|=` used before :D

Reply:
I haven't either! I used it hoping it would do what I thought it would do (intersection assignment)... and it did 😃